File: python-api/app/service/llm_service.py
import json
import logging
from typing import List
import httpx
from pydantic import BaseModel
from app.utils.settings import settings

LOG = logging.getLogger(__name__)


class LLMConfig(BaseModel):
    """Configuration for LLM model."""

    base_url: str
    model_name: str
    temperature: float
    timeout: int


class LLMService:
    """Service for interacting with Ollama LLM."""

    def __init__(self):
        self.config = LLMConfig(
            base_url=settings.ollama_base_url,
            model_name=settings.ollama_model_name,
            temperature=settings.ollama_temperature,
            timeout=settings.ollama_timeout,
        )
        self.client = httpx.AsyncClient(timeout=None)

    async def __aenter__(self):
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        await self.client.aclose()

    async def generate(self, prompt: str) -> str:
        """Generate raw text for a given prompt using LLM."""

        try:
            payload = {
                "model": self.config.model_name,
                "prompt": prompt,
                "stream": False,
                "options": {"temperature": self.config.temperature},
            }

            LOG.info(f"Requesting generation")
            LOG.info(f"Using Ollama at: {self.config.base_url}")
            LOG.info(f"Using model: {self.config.model_name}")

            response = await self.client.post(
                f"{self.config.base_url}/api/generate", json=payload
            )
            response.raise_for_status()

            result = response.json()
            llm_response = result.get("response", "")

            LOG.debug(f"LLM raw response: {llm_response}")

            return llm_response

        except httpx.ConnectError as e:
            LOG.error(f"Connection error to Ollama at {self.config.base_url}: {e}")
            raise Exception(
                f"Cannot connect to Ollama service at {self.config.base_url}. Please ensure Ollama is running."
            )
        except httpx.TimeoutException as e:
            LOG.error(f"Timeout error when calling Ollama: {e}")
            raise Exception(
                f"Timeout connecting to Ollama service (timeout: {self.config.timeout}s). Model might be loading or overloaded."
            )
        except httpx.HTTPError as e:
            LOG.error(f"HTTP error when calling Ollama at {self.config.base_url}: {e}")
            raise Exception(
                f"Failed to connect to Ollama service at {self.config.base_url}: {e}"
            )
        except Exception as e:
            LOG.error(f"Unexpected error generating llm: {e}")
            raise Exception(f"Failed to generate llm: {e}")

    async def generate_json(self, prompt: str) -> dict:
        """Generate json for a given prompt using LLM."""

        llm_response = await self.generate(prompt)

        # Parse the JSON response
        try:
            # Try to extract JSON from the response
            json_start = llm_response.find("{")
            json_end = llm_response.rfind("}") + 1

            if json_start != -1 and json_end > json_start:
                json_str = llm_response[json_start:json_end]
                json_data = json.loads(json_str)
            else:
                # Fallback: try to parse the entire response as JSON
                json_data = json.loads(llm_response)

            return json_data

        except json.JSONDecodeError as e:
            LOG.error(f"Failed to parse LLM JSON response: {e}")
            raise ValueError("Failed to parse LLM JSON response")

    async def health_check(self) -> bool:
        """Check if Ollama service is available."""
        try:
            response = await self.client.get(f"{self.config.base_url}/api/tags")
            return response.status_code == 200
        except Exception:
            return False

File: service/llm.service.go
package service

import (
	"bytes"
	"context"
	"disease-middleware/utils"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"regexp"

	log "github.com/sirupsen/logrus"
)

type OllamaRequest struct {
	Model   string                 `json:"model"`
	Prompt  string                 `json:"prompt"`
	Stream  bool                   `json:"stream"`
	Options map[string]interface{} `json:"options"`
}

type OllamaResponse struct {
	Response string `json:"response"`
}

type LLMService struct {
	client *http.Client
}

func NewLLMService() *LLMService {
	client := &http.Client{	}

	return &LLMService{
		client: client,
	}
}

func (s *LLMService) Generate(ctx context.Context, prompt string) (string, error) {
	log.Info("Requesting generation")
	log.Infof("Using Ollama at: %s", utils.Cfg.OllamaApi.Url)
	log.Infof("Using model: %s", utils.Cfg.OllamaApi.Model)

	payload := OllamaRequest{
		Model:  utils.Cfg.OllamaApi.Model,
		Prompt: prompt,
		Stream: false,
	}

	jsonData, err := json.Marshal(payload)
	if err != nil {
		log.Errorf("Error marshaling request: %v", err)
		return "", fmt.Errorf("failed to marshal request: %w", err)
	}

	url := fmt.Sprintf("%s/api/generate", utils.Cfg.OllamaApi.Url)
	req, err := http.NewRequestWithContext(ctx, "POST", url, bytes.NewBuffer(jsonData))
	if err != nil {
		log.Errorf("Error creating request: %v", err)
		return "", fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Content-Type", "application/json")

	resp, err := s.client.Do(req)
	if err != nil {
		log.Errorf("Connection error to Ollama: %v", err)
		return "", fmt.Errorf("cannot connect to Ollama service. Please ensure Ollama is running: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		log.Errorf("HTTP error when calling Ollama: status %d", resp.StatusCode)
		return "", fmt.Errorf("failed to connect to Ollama service: status %d", resp.StatusCode)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		log.Errorf("Error reading response body: %v", err)
		return "", fmt.Errorf("failed to read response: %w", err)
	}

	var result OllamaResponse
	if err := json.Unmarshal(body, &result); err != nil {
		log.Errorf("Error unmarshaling response: %v", err)
		return "", fmt.Errorf("failed to parse response: %w", err)
	}

	log.Debugf("LLM raw response: %s", result.Response)
	return result.Response, nil
}

func (s *LLMService) GenerateJSON(ctx context.Context, prompt string) (map[string]any, error) {
	llmResponse, err := s.Generate(ctx, prompt)
	if err != nil {
		return nil, err
	}

    // Try to extract JSON from the response using regex (from first "{" to last "}")
    re := regexp.MustCompile(`(?s)\{.*\}`)
    match := re.FindString(llmResponse)

    var jsonStr string
    if match != "" {
        jsonStr = match
    } else {
        // Fallback: try to parse the entire response as JSON
        jsonStr = llmResponse
    }

	var jsonData map[string]any
	if err := json.Unmarshal([]byte(jsonStr), &jsonData); err != nil {
		log.Errorf("Failed to parse LLM JSON response: %v", err)
		return nil, fmt.Errorf("failed to parse LLM JSON response: %w", err)
	}

	return jsonData, nil
}

func (s *LLMService) HealthCheck(ctx context.Context) bool {
	url := fmt.Sprintf("%s/api/tags", utils.Cfg.OllamaApi.Url)
	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
	if err != nil {
		return false
	}

	resp, err := s.client.Do(req)
	if err != nil {
		return false
	}
	defer resp.Body.Close()

	return resp.StatusCode == http.StatusOK
}
